### ISSUE TITLE
polling interface, like glutMainLoopEvent()

### ISSUE THREAD

**Issue by dobkeratops** [rikusalminen/glwt#8]
would you be open to exposing an alternative polling interface, the way glutMainLooEvent works http://freeglut.sourceforge.net/docs/api.php

This would be a good interface for my purposes,

```
{
  ... initialize, create twindows..
  while (1) {
     glwtProcessPendingMessages(); // same as 'glutMainLoopEvent'. handles UI changes
     update(dt);
     render(t);
     // update time..
 }
}
```

but even better would be something allowing more control:-

```
{
  //... initialize, create twindows..
  while (1) {
      while (glwtPeekMessage()) {    // 'glwtProcessPendingMessages()' is just this..
         GlwtWindowEvent event= glwtPopMessage();
         switch (event....) {
           // ... process the message ..
         }
      }
      update(dt,ctrl)
      render(t)
     // update time.
  }
}
```

// this is good because i can get accumulate information like controler input into local variables and pass to update. for the monolithic game view its convinient not to have to embed a pointer in a window, or go via global variables like in Glut. It would also be possible to use entirely without creating window callbacks if you wanted to.. a single game loop/view like on a console.

I understand android natively doesn't provide a polling interface. An implementation on android could do the threading behind the glwt interface to expose one

i like the fact glwt provides a big message enum/union, this will map well to rust's enums (just put the discriminant as the first struct element)

in Rust the above would look quite elegant..

```
{
  // ... create window..
  loop {
    while glwtPeekMessage() {
      match glwtPopMessage() {
        GlwtMouseMotion(win,(x,y),buttons)=> {
            //...
        }
        GlwtKeyDown(win, keysym,scancode,mod)=> {
           //...
        }
        // etc
        _=>
       }
    }
    update(..);
    render(..);
    // timer..
  }
}
```

if you had a 'GlwtEvent_None' you could turn that into one function call..  even simpler and less nesting

```
while match glwtPopMessage() {
    GlwtMouseMotion(win,(x,y),buttons)=>{ ...., true}
    GlwtKeyDown(win,keysym,scancode,mod)=>{ ...., true}
    GlwtEventNone=> { ....., false}
}
```

Do you provide a timer abstraction aswell? that would also be nice in this layer.

**Comment by progschj** [comment]
We had this discussion on the IRC channel a few times. We decided that the callback interface is best suited for GLWTs needs. But we are using only a single event struct exactly because that way a queue is easily implemented on the user end of the library (by just having a callback that pushes into a queue).

**Comment by jansol** [comment]
A timer abstraction was not originally planned but eventually we got to the consensus to have a glwtGetTime which would return time since an unspecified point in nanoseconds. However this proved unnecessarily complicated due to the way QueryPerformanceCounter works and nobody bothered to write a 128bit unsigned integer division so it has been stalled a bit. I have a version that uses doubles in one branch waiting to be either merged or scrapped. Anyway @rikusalminen might want to chime in on that.

**Comment by rikusalminen** [comment]
Thank you for the suggestion but this we have already discussed this quite a lot.

Ultimately, this is not a question about the interface but about the implementation details. The model you propose would require some kind of internal event queue. There's no one size fits all event queue model, so we decided against having an internal event queue in GLWT.

To summarize just some of the open ended questions in an event queue design... Should an event queue be thread safe? What kind of multithreaded semantics would it involve? Single consumer or several consumers? Should it be guarded by mutex/condition variable or should an unix pipe or something similar be used? Should the event queue have a finite capacity or re-allocate space as required? What should happen if the event queue is full? What if it is empty?

As far as I see the situation, there are no advantages to having an internal event queue in GLWT. As @progschj states above, it's very easy to create an event callback that puts the events in an event queue.

Additionally, some languages provide a concurrent channel primitive, Haskell and Go, for example. Even Rust has a channel type in the standard library (http://static.rust-lang.org/doc/0.9/std/comm/index.html). If GLWT was used in a language like that, it would make sense to use the channels that the language provides.

> Do you provide a timer abstraction aswell? that would also be nice in this layer.

There is a time branch that is created by @jansol. It's not yet merged to master, though. Ultimately we'd want to have a timer function that returns a monotonic timestamp in nanoseconds, but so far we're been using floating point numbers because converting from Win32 QueryPerformanceCounter/Frequency to nanoseconds would need 128 bit integer arithmetic.

Additionally, there has been some discussion about adding a timestamp to events, but they would be the timestamps associated with the events in the native windowing APIs and would represent the time when the event was _sent_ by the windowing system, not when it was _received_ by GLWT. Consequently, if there will be timestamps in events, they will not be using the same clock as the timer function.

In conclusion, it is very easy to create your own event queue and create a callback that puts events in the queue. And there's no advantage to having an internal event queue in GLWT.

I'll close this issue now but feel free to continue the discussion or ask any questions in this issue thread.
